File: utilities/nonvisualstudio/elixir/Spec/Primes.Spec.ex
# IN POWERSHELL
# iex.bat - REPL
# elixir c:\Users\Thomas\Documents\GitHub\practice\utilities\nonvisualstudio\elixir\Spec\Primes.Spec.ex

Code.require_file "../Implementation/Primes.ex", __DIR__

ExUnit.start

defmodule PrimesTests do
  use ExUnit.Case, async: true

  test "generatePrimes 14 should return [2,3,5,7,11,13]." do
    assert Primes.generatePrimes(14) == [2,3,5,7,11,13]
  end

  test "generatePrimes 11 should return [2,3,5,7,11]." do
    assert Primes.generatePrimes(11) == [2,3,5,7,11]
  end

  test "generatePrimes 1000 should return a list of length 168." do
    assert Primes.generatePrimes(1000) |> Enum.count == 168
  end

  test "generatePrimes 10000 should return a list of length 1229." do
    assert Primes.generatePrimes(10_000) |> Enum.count == 1229
  end

  test "whatExpondentOf2CreatesAnOddNumber 220 should return [2,55]." do
    assert Primes.whatExpondentOf2CreatesAnOddNumber(220) == [2, 55]
  end

  test "whatExpondentOf2CreatesAnOddNumber 210 should return [1, 105]." do
    assert Primes.whatExpondentOf2CreatesAnOddNumber(210) == [1, 105]
  end

  test "isProbablePrime 2 should return true." do
    assert Primes.isProbablePrime(2, 128) == true
  end

  test "isProbablePrime 1 should return false." do
    assert Primes.isProbablePrime(1, 128) == false
  end

  test "isProbablePrime -1 should return false." do
    assert Primes.isProbablePrime(-1, 128) == false
  end

  test "isProbablePrime 3 should return true." do
    assert Primes.isProbablePrime(3, 128) == true
  end

  test "isProbablePrime -3 should return true." do
    assert Primes.isProbablePrime(-3, 128) == true
  end

  test "isProbablePrime -6 should return false." do
    assert Primes.isProbablePrime(-6, 128) == false
  end

  test "isProbablePrime -9 should return false." do
    assert Primes.isProbablePrime(-9, 128) == false
  end

  test "isProbablePrime 961_748_932 should return false." do
    assert Primes.isProbablePrime(961_748_932, 128) == false
  end

  test "isProbablePrime 961_748_941 should return true." do
    assert Primes.isProbablePrime(961_748_941, 128) == true
  end

  test "isProbablePrime 961_748_931 should return false." do
    assert Primes.isProbablePrime(961_748_931, 128) == false
  end

  test "isProbablePrime 221 should return false." do
    assert Primes.isProbablePrime(221, 128) == false
  end

  test "isProbablePrime 5 should return true." do
    assert Primes.isProbablePrime(5, 128) == true
  end

  test "isProbablePrime 49_979_687 should return true." do
    assert Primes.isProbablePrime(49_979_687, 128) == true
  end

  test "isProbablePrime Math.pow(2,127) - 1 should return true." do
    lucasPrime = Math.pow(2,127) - 1
    assert Primes.isProbablePrime(lucasPrime, 128) == true
  end

  test "isProbablePrime Math.pow(2,127) - 3 should return false." do
    fakeLucasPrime = Math.pow(2,127) - 3
    assert Primes.isProbablePrime(fakeLucasPrime, 128) == false
  end

  test "isProbablePrime Math.pow(2, 148) + 1 / 17 should return true." do
    ferrierPrime = div((Math.pow(2, 148) + 1), 17)
    assert Primes.isProbablePrime(ferrierPrime, 128) == true
  end

  test "isProbablePrime Math.pow(2, 148) + 1 / 17 + 2 should return false." do
    ferrierPrime = div((Math.pow(2, 148) + 1), 17) + 2
    assert Primes.isProbablePrime(ferrierPrime, 128) == false
  end
end

File: utilities/nonvisualstudio/haskell/Spec/Primes.Spec.hs
{-
ghci c:\Users\Thomas\Documents\GitHub\practice\utilities\nonvisualstudio\haskell\Spec\Primes.Spec.hs
 c:\Users\Thomas\Documents\GitHub\practice\utilities\nonvisualstudio\haskell\Implementation\Primes.hs
 c:\Users\Thomas\Documents\GitHub\practice\utilities\nonvisualstudio\haskell\Implementation\Math.hs
-}
-- :r :q :set +s

module PrimesTests where

  import Test.HUnit
  import System.IO

  import Primes

  lazyPrimesTests = TestList
    [
      TestCase $ assertEqual "diffInfinite take 5 [1..] [2,4..] should return [1,3,5,7,9]." [1,3,5,7,9] (take 5 (diffInfinite [1..] [2,4..])),
      TestCase $ assertEqual "diffInfinite take 5 [10..] [2,4..])] should return [11,13,15,17,19]." [11,13,15,17,19] (take 5 $ diffInfinite [10..] [2,4..]),

      TestCase $ assertEqual "lazyPrimes take 5 should return [2,3,5,7,11]." [2,3,5,7,11] (take 5 lazyPrimes),
      TestCase $ assertEqual "lazyPrimes takeWhile (<10) should return [2,3,5,7]." [2,3,5,7] (takeWhile (<10) lazyPrimes),
      TestCase $ assertEqual "lazyPrimes takeWhile (<14) should return [2,3,5,7]." [2,3,5,7,11,13] (takeWhile (<14) lazyPrimes),

      TestCase $ assertEqual "randomNumber should not repeat with very large number" False ((randomNumber (1,1239009123984198)) == (randomNumber (1,1239009123984198))),

      TestCase $ assertEqual "whatExpondentOf2CreatesAnOddNumber 220 should return (2,55)." (2,55) (whatExpondentOf2CreatesAnOddNumber 220 (0,220)),
      TestCase $ assertEqual "whatExpondentOf2CreatesAnOddNumber 210 should return (1,105)." (1,105) (whatExpondentOf2CreatesAnOddNumber 210 (0,210)),

      --TestCase $ assertEqual "isProbablePrime 1 should return False." False (isProbablePrime 1 128),
      --TestCase $ assertEqual "isProbablePrime -1 should return False." False (isProbablePrime (-1) 128),
      --TestCase $ assertEqual "isProbablePrime 0 should return False." False (isProbablePrime 0 128),
      --TestCase $ assertEqual "isProbablePrime -6 should return False." False (isProbablePrime (-6) 128),
      --TestCase $ assertEqual "isProbablePrime -9 should return False." False (isProbablePrime (-9) 128),
      --TestCase $ assertEqual "isProbablePrime 961748932 should return False." False (isProbablePrime 961748932 128),
      --TestCase $ assertEqual "isProbablePrime 961748931 should return False." False (isProbablePrime 961748931 128),
      --TestCase $ assertEqual "isProbablePrime 221 should return False." False (isProbablePrime 221 128),
      --TestCase $ assertEqual "isProbablePrime 3561 should return False." False (isProbablePrime 3561 128),
      TestCase $ assertEqual "isProbablePrime ((2 ^ 127) - 3) should return False." False (isProbablePrime ((2 ^ 127) - 3) 128),
      TestCase $ assertEqual "isProbablePrime ((div ((2 ^ 148) + 1) 17) + 2) should return False." False (isProbablePrime ((div ((2 ^ 148) + 1) 17) + 2) 128),

      --TestCase $ assertEqual "isProbablePrime 3571 should return True." True (isProbablePrime 3571 128),
      --TestCase $ assertEqual "isProbablePrime 2 should return True." True (isProbablePrime 2 128),
      --TestCase $ assertEqual "isProbablePrime 3 should return True." True (isProbablePrime 3 128),
      --TestCase $ assertEqual "isProbablePrime -3 should return True." True (isProbablePrime (-3) 128),
      --TestCase $ assertEqual "isProbablePrime 961748941 should return True." True (isProbablePrime 961748941 128),
      TestCase $ assertEqual "isProbablePrime 49979687 should return True." True (isProbablePrime 49979687 128),
      TestCase $ assertEqual "isProbablePrime ((2 ^ 127) - 1) should return True." True (isProbablePrime ((2 ^ 127) - 1) 128),
      TestCase $ assertEqual "isProbablePrime div ((2 ^ 148) + 1) 17 should return True." True (isProbablePrime (div ((2 ^ 148) + 1) 17) 128)

    ]

  testCases = TestList
    [
      lazyPrimesTests
    ]

  tests = runTestTT testCases
